linux/GPU.c

```diff
@@ -0,0 +1,272 @@
+/*
+htop - GPU.c
+(C) 2023 htop dev team
+Released under the GNU GPLv2+, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include "linux/GPU.h"
+
+#include <assert.h>
+#include <ctype.h>
+#include <dirent.h>
+#include <errno.h>
+#include <sys/types.h>
+
+#include "XUtils.h"
+
+#include "linux/LinuxMachine.h"
+
+
+#define INVALID_CLIENT_ID ((unsigned long long int)-1)
+
+
+struct client_id {
+   char* pdev;
+   unsigned long long int id;
+   struct client_id* next;
+};
+
+enum section_state {
+   SECST_UNKNOWN,
+   SECST_DUPLICATE,
+   SECST_NEW,
+};
+
+static bool null_str_eq(const char* s1, const char* s2) {
+   if (s1 == NULL && s2 == NULL)
+      return true;
+
+   if (s1 && s2)
+      return String_eq(s1, s2);
+
+   return false;
+}
+
+static bool is_duplicate_client(const struct client_id* parsed, unsigned long long int id, const char* pdev) {
+   for (; parsed; parsed = parsed->next) {
+      if (id == parsed->id && null_str_eq(pdev, parsed->pdev)) {
+         return true;
+      }
+   }
+
+   return false;
+}
+
+static void update_system(LinuxProcessList* lpl, unsigned long long int time, char* engine) {
+   Machine* host = lpl->super.host;
+   LinuxMachine* lhost = (LinuxMachine*) host;
+   GPUEngineData** engineData = &lhost->gpuEngineData;
+
+   while (*engineData) {
+      if (String_eq((*engineData)->key, engine))
+         break;
+
+      engineData = &((*engineData)->next);
+   }
+
+   if (!*engineData) {
+      GPUEngineData* newData;
+
+      newData = xMalloc(sizeof(*newData));
```

Comment:
```suggestion
      GPUEngineData* newData = xMalloc(sizeof(*newData));
```

Reply:
Or better:
```suggestion
      *engineData = xMalloc(sizeof(**engineData));
```

Reply:
The current code is more readable for me, e.g. it avoids ` **engineData = ...`